contracts/contracts/buyback/Buyback.sol

```diff
@@ -52,26 +68,78 @@ contract Buyback is Governable {
      * protocol (e.g. Sushiswap)
      **/
     function swap() external onlyVault {
+        // Don't revert everything, even if the buyback fails.
+        // We want the overall transaction to continue regardless.
+        (bool success, bytes memory data) = address(this).call(
+            abi.encodeWithSignature("swapAndCheck()")
+        );
+        if (!success) {
+            emit BuybackFailed(data);
+        }
+    }
+
+    function swapAndCheck() external {
+        require(
+            msg.sender == address(this),
+            "Buyback: swapAndCheck only internal"
+        );
         if (uniswapAddr == address(0)) return;
 
         uint256 sourceAmount = ousd.balanceOf(address(this));
         if (sourceAmount < 1000 * 1e18) return;
 
-        // Uniswap redemption path
-        address[] memory path = new address[](4);
-        path[0] = address(ousd);
-        path[1] = address(usdt);
-        path[2] = IUniswapV2Router(uniswapAddr).WETH();
-        path[3] = address(ogn);
-        IUniswapV2Router(uniswapAddr).swapExactTokensForTokens(
-            sourceAmount,
-            uint256(0),
-            path,
-            address(this),
-            now
+        UniswapV3Router.ExactInputParams memory params = UniswapV3Router
+            .ExactInputParams({
+            path: abi.encodePacked(
+                ousd,
+                uint24(500), // Pool fee, ousd -> usdt
+                usdt,
+                uint24(3000), // Pool fee, usdt -> weth9
+                weth9,
+                uint24(3000), // Pool fee, weth9 -> ogn
+                ogn
+            ),
+            recipient: address(this),
+            deadline: uint256(block.timestamp + 1000),
+            amountIn: sourceAmount,
+            amountOutMinimum: uint256(0)
```

Comment:
Is this safe?  If one of those three pools gets extremely lopsided somehow, we could end up with a poor `amountOut`.  Wonder if we should add a guesstimate here to prevent something like that?

Reply:
Checked after the TX, but yes, might just change this to run here.

Reply:
Done